odoo/orm/domains.py

```diff
@@ -350,39 +354,50 @@ def map_conditions(self, function: Callable[[DomainCondition], Domain]) -> Domai
     def validate(self, model: BaseModel) -> None:
         """Validates that the current domain is correct or raises an exception"""
         # just execute the optimization code that goes through all the fields
-        self._optimize_for_sql(model)
+        self.optimize(model, full=True)
 
-    def _optimize(self, model: BaseModel) -> Domain:
+    def optimize(self, model: BaseModel, *, full: bool = False) -> Domain:
         """Perform optimizations of the node given a model.
 
-        The optimizations are essentially model agnostic and transaction
-        independent. This means that they only depend on the model's fields
-        definitions. No model-specific override is used, and the resulting
-        domain may be reused in another transaction without semantic impact.
-        The model's fields are used to validate conditions and apply
-        type-dependent optimizations.
-
-        This optimization level may be useful to simplify a domain that is sent
-        to the client-side, thereby reducing its payload/complexity.
-
         It is a pre-processing step to rewrite the domain into a logically
         equivalent domain that is a more canonical representation of the
         predicate. Multiple conditions can be merged together.
+
+        By default, ``full=False`` applies basic optimizations only. Those are
+        transaction-independent; they only depend on the model's fields
+        definitions. No model-specific override is used, and the resulting
+        domain may be reused in another transaction without semantic impact.
+        The model's fields are used to validate conditions and apply
+        type-dependent optimizations. This optimization level may be useful to
+        simplify a domain that is sent to the client-side, thereby reducing its
+        payload/complexity.
+
+        With ``full=True``, basic and advanced optimizations are applied.
+        Additional optimizations may rely on model specific overrides
+        (search methods of fields, etc.) and the semantic equivalence is only
+        guaranteed at the given point in a transaction. We resolve inherited
+        and non-stored fields (using their search method) to transform the
+        conditions.
         """
-        return self
+        level = OptimizationLevel.FULL if full else OptimizationLevel.BASIC
+        if self._opt_level >= level:
+            return self
 
-    def _optimize_for_sql(self, model: BaseModel) -> Domain:
-        """Perform optimizations required by `_to_sql`.
+        # determine a fixpoint for _optimize()
+        previous, domain, count = None, self, 0
+        while domain != previous:
```

Comment:
Since you add a lot of condition like this. Is it worth it to have a shortcut in `__eq__` `(self is other or (...)) ` to avoid comparing a lot of object ?

Reply:
Most are quite simple, it would make sense only for `DomainNary` (where the list could be long), and maybe for `DomainCondition` (because of value).